### PR TITLE
Lua string compress

### PR DESCRIPTION
**Pull request by NiLSPACE**
Exported CompressString and UncompressString to the Lua API

### CODE REVIEW

**Comment by NiLSPACE** [comment]
I noticed there are more functions like CompressStringGZIP, UncompressStringGZIP and InflateString. Should I export those as well?

**Comment by madmaxoft** [comment]
These bindings are not good. they should not have the `Length` parameter, because that one is implicit in Lua strings (and C++ `std::string`).

You should use something like this:
```cpp
static int tolua_CompressString(lua_State * tolua_S)
{
  cLuaState S(tolua_S);
  // TODO: Check param types

  // Get the params:
  AString ToCompress;
  int CompressionLevel;
  S.GetStackValues(1, ToCompress, CompressionLevel);

  // Compress the string:
  AString res;
  CompressString(ToCompress.data(), ToCompress.size(), res, CompressionLevel);
  S.Push(res);
  return 1;
}
```

Then the plugins can use code like this:
```lua
local FileContents = cFile:ReadWholeFile(...)
local Compressed = CompressString(FileContents)
```

**Comment by NiLSPACE** [comment]
I asume from looking at the code that there is an optional second parameter for the compressionlevel?

But how would the uncompress function look like? Would that look like this?:
```lua
UncompressString(a_CompressedString, a_UncompressedSize)
```

**Comment by madmaxoft** [comment]
Yes, unfortunately the compression algorithm doesn't store the uncompressed length, so it needs it as an input.

How about putting all those functions into a `cStringCompression` namespace (similar to `cCryptoHash` namespace)? We are likely to add a few more in the future, so it would make sense.

**Comment by NiLSPACE** [comment]
Good idea

**Comment by NiLSPACE** [comment]
Also, how about we rename the current exported functions to CompressStringZLIB so they match the GZIP functions?

**Comment by madmaxoft** [comment]
Makes sense.

**Comment by NiLSPACE** [comment]
How is this?

**Comment by madmaxoft** [comment]
We've got `CheckParamXYZ()` functions in `cLuaState` that check the param type and output an error if the param doesn't match, you should use those. Right now an erroneous parameter will cause a stack trace with no info whatsoever.

**Comment by NiLSPACE** [comment]
Like this?

File: src/Bindings/ManualBindings.cpp [hunk 1]
```diff
@@ -110,6 +111,146 @@ static int tolua_Clamp(lua_State * tolua_S)
...
 
 
+static int tolua_CompressStringZLIB(lua_State * tolua_S)
+{
+	cLuaState S(tolua_S);
+	if (
+		!S.CheckParamString(1) &&
+		(
+		!S.CheckParamNumber(2) ||
+		!S.CheckParamEnd(2)
```
Comment (madmaxoft) [inline src/Bindings/ManualBindings.cpp]: Extra indent should go here.

**Comment by madmaxoft** [comment]
I'm wondering, don't we have the opposite function to Inflate? Makes no sense to have just one direction.

File: src/Bindings/ManualBindings.cpp [hunk 2]
```diff
@@ -110,6 +111,146 @@ static int tolua_Clamp(lua_State * tolua_S)
 
 
 
+static int tolua_CompressStringZLIB(lua_State * tolua_S)
+{
+	cLuaState S(tolua_S);
+	if (
+		!S.CheckParamString(1) &&
+		(
+			!S.CheckParamNumber(2) ||
```
Comment (NiLSPACE) [inline src/Bindings/ManualBindings.cpp]: I'm confusing myself. Should that really be OR?

Comment (madmaxoft) [inline src/Bindings/ManualBindings.cpp]: Nope, this should be an AND, and all the others should be ORs. You want to fail if any of the top-level tests fail.

**Comment by madmaxoft** [comment]
And that's why you should write a testing plugin for each new API that you make.

**Comment by NiLSPACE** [comment]
I usualy test new API's using my Execute plugin

**Comment by NiLSPACE** [comment]
The build failed because the Google test failed.

**Comment by NiLSPACE** [comment]
Again the google test failed

**Comment by NiLSPACE** [comment]
Hmm. Currently there is no way to know if the compression failed or succeeded. Maybe we should add another return value?

**Comment by madmaxoft** [comment]
If the compression fails, the returned string is empty. Similar with decompression.

**Comment by NiLSPACE** [comment]
So is this mergeable?

**Pull request merged**